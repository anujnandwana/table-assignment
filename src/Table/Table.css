
#data-table-wrapper{
  margin:10px;
  border: 1px solid #dedede;
}

.data-table{
  width: 100%;
  border-collapse: collapse;
}
.data-table tr{
  border: 1px solid #dedede;
}
.data-table th::first-letter,.data-table .status::first-letter{
  text-transform: uppercase;
}

.data-table td,.data-table th{
  text-align: left;
  padding: 12px;
}

.data-table caption button{
  border: 0;
  color: #a3a3a3;
  background: white;
  font-size: 20px;
  cursor: not-allowed;
  pointer-events:none;
}

.data-table caption button.active{
  color: #000;
  cursor: pointer;
  pointer-events:unset;
}

.data-table .status[data-status="available"]::before{
  content: '';
  position: absolute;
  transform: translate(-20px, 50%);
  width: 10px;
  height: 10px;
  background-color: green;
  border-radius: 50%;
}

caption{
  padding: 15px;
  font-size: 20px;
  text-align: left;
}

caption .selected-text{
  margin-left: 4vw;
  margin-right: 4vw;
}

input[type="checkbox"]{
  width: 20px; 
  height: 20px;
}

.data-table tbody tr:hover{
  background-color: #f7f7f7;
}
.data-table tbody tr.active{
  background-color: #e9e9e9;
}
