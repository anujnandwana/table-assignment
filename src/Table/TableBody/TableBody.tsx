import React from "react";
import { TableData } from "../../data/data";
import TableRow from "../TableRow/TableRow";

interface TableBodyProps {
  tableData: TableData[];
  tableHeadings: string[];
  selectedRowIds: string[];
  setSelectedRowIds: (ids: string[]) => void;
}

const TableBody: React.FC<TableBodyProps> = ({ tableData, tableHeadings, selectedRowIds, setSelectedRowIds }) => {

  /**
   * Handles the selection and deselection of table rows
   * @param isSelected - Rows checkbox is selected or not
   * @param selectedId - The id associated with the checkbox
   */
  const handleSelection = function (isSelected: boolean, selectedId: string) {

    if (isSelected) {
      setSelectedRowIds([...selectedRowIds, selectedId]);
    } else {
      const selectedIds: string[] = [];
      selectedRowIds.forEach((id) => {
        if (id !== selectedId) {
          selectedIds.push(id);
        }
      });
      setSelectedRowIds(selectedIds);
    }

  };

  return (
    <tbody className="data-table-container">
      {tableData.map((rowData) => {
        return (
          <TableRow
            key={rowData.id}
            rowData={rowData}
            tableHeadings={tableHeadings}
            isSelected={selectedRowIds.includes(rowData.id as string)}
            onSelectRow={handleSelection}
          />
        );
      })}
    </tbody>
  );
};

export default TableBody;