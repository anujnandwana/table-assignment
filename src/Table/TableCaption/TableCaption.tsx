import React from "react";
import { TableData } from "../../data/data";
import DownloadAction from "../Actions/Download";
import SelectAll from "../Actions/SelectAll";
import SelectedText from "../Actions/SelectedText";

interface TableCaptionProps {
  selectedRowIds: string[];
  setSelectedRowIds: (ids: string[]) => void;
  selectAll: boolean;
  tableData: TableData[];
}

const TableCaption: React.FC<TableCaptionProps> = ({
  selectedRowIds,
  setSelectedRowIds,
  selectAll,
  tableData,
}) => {
  const [downloadAbleData, setDownloadAbleData] = React.useState<TableData[]>([]);

  
  /**
   * Update the disable/enable status of download btn by checking available status
   */
  React.useEffect(() => {
    const availableStatusData = tableData.filter(
      (row) =>
        selectedRowIds.includes(row.id as string) && row.status === "available"
    );
    setDownloadAbleData(availableStatusData);
  }, [tableData, selectedRowIds]);


  return (
    <caption>
      <SelectAll selectAll={selectAll} setSelectedRowIds={setSelectedRowIds} tableData={tableData} />
      <SelectedText selectedRowIds={selectedRowIds} />
      <DownloadAction downloadAbleData={downloadAbleData} />
    </caption>
  );
};

export default TableCaption;
