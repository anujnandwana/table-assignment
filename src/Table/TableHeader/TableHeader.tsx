import React from "react";

interface TableHeaderProps {
    tableHeadings: string[];
}

const TableHeader: React.FC<TableHeaderProps> = ({ tableHeadings }) => {
    return (
        <thead>
            <tr>
                {tableHeadings.map((heading, headIndex) => {
                    return (
                        <th key={`head-${headIndex}`}>
                            {heading === "id" ? "" : heading}
                        </th>
                    );
                })}
            </tr>
        </thead>
    );
};

export default TableHeader;