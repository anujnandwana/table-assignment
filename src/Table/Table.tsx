import React from "react";
import { TableData } from "../data/data";
import "./Table.css";
import TableBody from "./TableBody/TableBody";
import TableCaption from "./TableCaption/TableCaption";
import TableHeader from "./TableHeader/TableHeader";
import { v4 as uuidv4 } from 'uuid';

interface TableProps {
  data: TableData[];
}

function Table(props: TableProps) {
  const [tableData, setTableData] = React.useState<TableData[]>([]);
  const [selectedRowIds, setSelectedRowIds] = React.useState<string[]>([]);
  const [selectAll, setSelectAll] = React.useState(false);


  React.useEffect(() => {
    const tableIdsData: TableData[] = props?.data.map((rowData) => {
      // Adding the uniqueId to each row  
      rowData = { id: uuidv4(), ...rowData };
      return rowData;
    });

    setTableData(tableIdsData);
  }, [props?.data]);


  /** 
   * Handles the selectAll checkbox state by checking if all the checkbox is selected or not
   */
  React.useEffect(() => {
    setSelectAll(selectedRowIds.length === tableData.length)
  }, [selectedRowIds, tableData]);


  /**
   * Return the Array of table headings.
   * Memoize the calculation of table headings 
   * to avoid recomputation on each render for performance.
   */
  const headings = React.useMemo(() => {
    const rowHeadings = tableData.reduce((acc: string[], curr: Object) => {
      acc.push(...Object.keys(curr));
      return acc;
    }, []);
    return Array.from(new Set(rowHeadings));
  }, [tableData]);


  return (
    <div id={"data-table-wrapper"}>
      <table className="data-table" role="grid" aria-label="Data Table">
        <TableCaption selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} tableData={tableData} selectAll={selectAll} />
        <TableHeader tableHeadings={headings} />
        <TableBody tableData={tableData} tableHeadings={headings} selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} />
      </table>
    </div>
  );
}

export default Table;
