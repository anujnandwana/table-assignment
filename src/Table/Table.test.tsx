import Table from './Table';
import { mount } from 'enzyme';

import { TableData } from '../data/data';

import TableCaption from "../Table/TableCaption/TableCaption"
import TableBody from './TableBody/TableBody';
import TableHeader from './TableHeader/TableHeader';

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'mocked-id'),
}));


describe('Table component', () => {
  const mockData: TableData[] = [
    { device: 'Device 1', path: '/path/1', name: 'jui', status: 'available' }
  ];

  test('Table component should match snapshot', () => {
    const component = mount(<Table data={mockData} />);
    expect(component).toMatchSnapshot();
  });

  test('Selected component should have no selected id if not selected ', () => {
    const component = mount(<Table data={mockData} />);
    const tableCaption = component.find(TableCaption);

    expect(tableCaption.length).toBe(1)
    expect(tableCaption.at(0).prop('selectedRowIds').length).toBe(0)
  });

  test('TableHeader should have proper header according to the data with unique Id', () => {
    const component = mount(<Table data={mockData} />);
    const tableHeader = component.find(TableHeader);

    const updatedData = ["id", ...Object.keys(mockData[0])]


    expect(tableHeader.length).toBe(1)
    expect(tableHeader.at(0).prop('tableHeadings')).toStrictEqual(updatedData)
  });
});
