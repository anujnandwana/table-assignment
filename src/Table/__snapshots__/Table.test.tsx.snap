// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component Table component should match snapshot 1`] = `ReactWrapper {}`;
